### ISSUE TITLE
Provide SARIF output

### ISSUE THREAD

**Issue by chris48s** [chris48s/v8r#148]
Aside from the exit code, all of v8r's output is optimised for reading by humans. It would be useful to also provide some kind of machine-readable output (e.g: via a `--output json` flag, or something)

In #21 @nvuillam suggested SARIF format:

> what about SARIF output ? https://sarifweb.azurewebsites.net/
>
> More and more analyzers use this common format, and to be honest it would help me to integrate v8r even better within MegaLinter :)

> eslint has a SARIF formatter, already widely used :) ( 50k dl / week )
>
> https://www.npmjs.com/package/@microsoft/eslint-formatter-sarif

> Here is an example: there are some standard attributes (not sure any of them is required), and for anything not in the spec you can put it in `properties`

 <details>
  <summary>SARIF example</summary>
<pre>{
  "runs": [
    {
      "tool": {
        "driver": {
          "name": "Bandit",
          "rules": [
            {
              "id": "B110",
              "name": "try_except_pass",
              "helpUri": "https://bandit.readthedocs.io/en/latest/plugins/b110_try_except_pass.html"
            }
          ]
        }
      },
      "invocations": [
        {
          "executionSuccessful": true,
          "endTimeUtc": "2021-12-05T00:20:06Z"
        }
      ],
      "results": [
        {
          "message": {
            "text": "Try, Except, Pass detected."
          },
          "level": "note",
          "locations": [
            {
              "physicalLocation": {
                "region": {
                  "snippet": {
                    "text": "except:\n"
                  },
                  "startLine": 3
                },
                "artifactLocation": {
                  "uri": "file:///github/workspace/python_bad_1.py"
                },
                "contextRegion": {
                  "snippet": {
                    "text": "    pass\nexcept:\n    pass\n"
                  },
                  "endLine": 4,
                  "startLine": 2
                }
              }
            }
          ],
          "properties": {
            "issue_confidence": "HIGH",
            "issue_severity": "LOW"
          },
          "ruleId": "B110",
          "ruleIndex": 0
        }
      ]
    }
  ],
  "version": "2.1.0",
  "$schema": "https://schemastore.azurewebsites.net/schemas/json/sarif-2.1.0-rtm.4.json"
}</pre>
</details> 

Would this be a good fit for v8r?

**Comment by nvuillam** [comment]
@chris48s FYI I'm currently building a SarifBuilder module package (that I'll use within npm-grovy-lint that requires it ^^)

It will allow to simply build SARIF reports without studying the whole OASIS norm (but also allow purists to build extended ones), so if you were about to start SARIF output for v8r, it may be faster if you wait for my package to be ready :)

**Comment by nvuillam** [comment]
@chris48s , FYI [node-sarif-builder](https://github.com/nvuillam/node-sarif-builder) is available and documented :)

I succeeded to plug it within npm-groovy-lint quite easily , should be ok with v8r too ^^

**Comment by chris48s** [comment]
In order for this to be machine-readable the user will need to be able to suppress any other logging output as well as outputting JSON.

Some tools (trivy springs to mind) output log statements as well as JSON by default and you have to suppress it with `--quiet`

Another approach would be to auto-suppress other log output when outputting JSON/SARIF format.

**Comment by chris48s** [comment]
I've been having having a bit of a noodle on this.

Having looked at it, there are a few places where I am struggling to bash this square peg into a SARIF-shaped hole which stem from the fact that ajv (and by extension v8r) is a validator, not a code quality tool. They are closely related but not the same thing.

One issue I have is that what we are really validating here is a parsed object rather than a source file. So for example the following source files

```json
{ "foo": { "bar": 5 } }
```

```json
{
  "foo": {
    "bar": 5
  }
}
```

```yaml
# yaml
foo:
  bar: 5
```

would all parse to the same object. Then when we validate it, ajv will express this as "there's a problem with the `bar` property" rather than "there's an error on line 4" (or whatever). This is fine, but as far as I can tell, the only way that SARIF allows us to specify the 'location' of an error is `startLine` / `endLine`.

The other issue is that there are fairly important things I'd like to communicate as structured data in the machine readable output like "did the file pass or fail validation: yes/no" or "what schema was the file validated against?" that don't map neatly to SARIF constructs. They can be communicated in a `"note"` result, or included in a property bag but there's definitely a bit of a mismatch of objectives between the result of a validation (pass/fail) and the result of a code quality tool (here's some issues that might require attention).

As much as I'd prefer to lean on an existing standard where possible for interoperability I think I'm leaning towards the main or default JSON output being based on AJV's output with a few additional properties rather than SARIF. I think if I do produce SARIF it would be a secondary option. If I were to shoehorn the result into a less useful but SARIF-formatted output (that doesn't include any line numbers for the errors) as a secondary output, is that actually useful? I don't have the desire to take on maintenance of a feature that doesn't really meet the needs of the one person who would actually care about using it.

I'm slowly churning this one round..

**Comment by nvuillam** [comment]
So I can assure you that if you implement SARIF, that won't be just for MegaLinter use, as it became the community format for analysis results ^^

**Comment by chris48s** [comment]
I think the question I'm interested in is not so much "is SARIF a popular format?" but more like "what is the minimum subset of SARIF required for output to be useful to you?"

For example, you mention 

> (ex: Github that can display errors directly in files in their only pull request UIs)

but if you look at the documentation for the minimum subset of SARIF that integration requires https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning there are definitely fields which are required there that I wouldn't currently be able to populate.

I'm sure I can output some blob of JSON which is valid SARIF but it it doesn't actually communicate information that the one person who currently wants it needs to consume, its a pointless endeavour. Lets not worry about potential future consumers - right now they represent a fictional use-case.

> and eventually property name

What construct in the SARIF spec would appropriate to use for communicating the error path? (AJV provides this as a JSON Pointer)

**Comment by chris48s** [comment]
Some other notes:

> Some tools (trivy springs to mind) output log statements as well as JSON by default and you have to suppress it with `--quiet`
> 
> Another approach would be to auto-suppress other log output when outputting JSON/SARIF format.

Another solution to this would be to send machine-readable output to stdout and human-readable log messages to stderr. I think I like this one the most.

---

Conceptually I feel like SARIF output might be best delivered as a separate plugin as opposed to a core feature of v8r, but I wonder if the scaffolding needed to allow it to be a separate plugin is more of a maintenance overhead than the feature itself. It remains an interesting rabbit hole though.

**Comment by nvuillam** [comment]
I think that incomplete SARIF is still better than no sarif at all ^^

**Comment by chris48s** [comment]
Sure, but there must be some minimum set of fields megalinter needs/expects for a valid integration.

If you can answer that question it is likely you will get what you want. It seems wierd to be avoiding the question.

**Comment by chris48s** [comment]
I've now shipped [v0.13.0](https://www.npmjs.com/package/v8r/v/0.13.0) which outputs text/json to stdout and log messages to stderr. Now changing the title of this issue to reflect that the remaining conversation is only about trying to nail down approach to SARIF and add that format.

**Comment by chris48s** [comment]
Hello.

I've just released v8r 4.0.0.

This release introduces a feature relevant to this issue: It is now possible to extend the functionality of v8r by installing or writing plugins. This includes adding alternate output formats, so would be a good fit for this. There is documentation on the plugin system at https://chris48s.github.io/v8r/category/plugins/

I've taken the start I made on a SARIF implementation in https://github.com/chris48s/v8r/pull/217 and re-written that as a v8r plugin compatible with v4 here:

```js
import { createRequire } from "module";
// TODO: once JSON modules is stable these requires could become imports
// https://nodejs.org/api/esm.html#esm_experimental_json_modules
const require = createRequire(import.meta.url);

import path from "path";
import url from "url";

import {
  SarifBuilder,
  SarifRunBuilder,
  SarifResultBuilder,
} from "node-sarif-builder";

import { BasePlugin } from "v8r";

class SarifOutput extends BasePlugin {
  static name = "v8r-plugin-sarif-output";

  registerOutputFormats() {
    return ["sarif"];
  }

  getAllResultsLogMessage(results, format) {
    if (format === "sarif") {
      const v8rMain = require.resolve("v8r");
      const pkg = require(
        path.join(path.parse(v8rMain).dir, "..", "package.json"),
      );

      const sarifBuilder = new SarifBuilder();
      const sarifRunBuilder = new SarifRunBuilder().initSimple({
        toolDriverName: pkg.name,
        toolDriverVersion: pkg.version,
        url: pkg.homepage,
      });
      for (const result of results) {
        if (result.valid) {
          const sarifResultBuilder = new SarifResultBuilder();
          const sarifResultInit = {
            level: "note",
            messageText: "file is valid",
            fileUri: url.pathToFileURL(result.fileLocation).href,
          };
          sarifResultBuilder.initSimple(sarifResultInit);
          sarifRunBuilder.addResult(sarifResultBuilder);
          continue;
        }

        for (const error of result.errors) {
          const sarifResultBuilder = new SarifResultBuilder();
          const sarifResultInit = {
            level: "error",
            ruleId: error.keyword,
            messageText: error.message,
            fileUri: url.pathToFileURL(result.fileLocation).href,
          };
          sarifResultBuilder.initSimple(sarifResultInit);
          sarifRunBuilder.addResult(sarifResultBuilder);
        }
      }
      sarifBuilder.addRun(sarifRunBuilder);
      const sarifJsonString = sarifBuilder.buildSarifJsonString({
        indent: true,
      });
      return sarifJsonString;
    }
  }
}

export default SarifOutput;
```

That gives you a starting point (or maybe even a good enough implementation? I was never really clear) for a v8r SARIF plugin, although the limitation that we're validating a parsed object rather than a source file stands.

I'm going to close this issue as the plugin system now gives you a way to implement this if you want. That said, if you take this forward I'm happy to field follow up questions if you have them.
